### ISSUE TITLE
Linux Port

### ISSUE THREAD

**Issue by samosaara** [LazoCoder/Pokemon-Terminal#7]
What would be needed to re-do/change in order for a Linux port? How much is the current 'method' tied to iTerm2/macOSx? Is this an iTerm2 feature?

Because AFAIK, images on Linux terminals are a deadly sin, except maybe for terminology... So, what kind of work needs doing till I can add this to my rice :smiley:

**Comment by BnMcG** [comment]
After taking a look at the code, it seems like this works by generating AppleScript from the user's choice, and then executing it. I've never really dug into changing many Linux desktop environment features programatically, but I would assume you could add to the Scripter python file to generate some Bash script to change the desktop background (and possibly the terminal profile to change a background image?), and then pick which generated code to run based on the platform.

**Comment by samosaara** [comment]
So, this is a two part thing then, the desktop and the terminal background, in Linux almost absolutely everything is interchangeable and sadly not always compatible so for the desktop background part we will need to add a different method for each DE (desktop envoirment) which I bet would very simple actually, just a little bit annoying because we would need to add a different line of code to each different one, but there is a problem with terminal emulator.

Linux has way more terminal emulators, and ironically almost none of them supports image (TTY's supports images which is even more ironic) the only ones that **I know** do, is terminology and hyper.  Terminology is actually the closest thing we have to iTerm2 actually. Hyper is simply a electron CSS/HTML/JS based terminal so, well, CSS.

So to sum it up: to the desktop background part, it is needed to refactor the bash scripts to check which DE is running and call the appropriate background setting script (if supported). And same for terminal emulator? So the python, scripts are compatible, they only choose and pass forward the image file?

**Comment by LazoCoder** [comment]
I think I did a pretty good job at modularizing the code so I don't think you need to change anything outside of scripter.py to get this to work.
@BnMcG is correct, it works by generating AppleScript. If you want to make a Linux port the only thing you need to understand is the following methods in scripter.py:
```def change_terminal(pokemon)```
and
```def change_wallpaper(pokemon)```
As you can see, for example, the method *change_terminal(pokemon)* calls the following method, which essentially generates a string containing the entire script. It is the meat of the project:
```def __terminal_script(pokemon)```
If you are adding a Linux port then this method should probably be renamed to something more specific:
```def __iTerm2_terminal_script(pokemon)```
Or something like that. And then you would make a different method for each linux distro / terminal that you want. Maybe something like:
```
def __terminology_terminal_script(pokemon)
def __hyper_terminal_script(pokemon)
# etc...
```
Each method should generate the script to change the terminal as per that particular environment (don't forget about wallpaper changing too!). Then you can just adjust the method *change_terminal(pokemon)* to pick which script to call and execute.
I was going to write the linux port but I have work and I need to install ubuntu again so I probably won't be able to get started on this until the weekend. If someone can get it done before then and it works, I'll make sure to accept the pull request.

**Comment by DuIslingr** [comment]
Could you port it to Hyper Terminal? https://github.com/zeit/hyper still somewhat relevant to linux since its cross platform and i use it on linux :p

**Comment by DRogue1337** [comment]
Maybe port it to every existing terminal emulator that supports background images? At least the most common ones. Since @LazoCoder did a good job modularizing you'd just need an extra function per special needs of the terminal and use that.

**Comment by samosaara** [comment]
I finished adding support for Linux GNOME and terminology, needs some testing on other distros, especially on more stable ones, since I did all the testing on an arch linux. I made sure to make it as modular as possible to support other WM's and image-capable terminal emulators. I think this issue should be closed after and if, the PR gets merged. and more specific issues for other WM's and terminal emulators to be open on-demand.
